src/nimble.nim

```diff
@@ -356,7 +368,9 @@ proc installFromDir(dir: string, requestedVer: VersionRange, options: Options,
   # if the build fails then the old package will still be installed.
   if pkgInfo.bin.len > 0:
     let paths = result.deps.map(dep => dep.getRealDir())
-    buildFromDir(pkgInfo, paths, true)
+    var optsCopy = options.briefClone()
+    optsCopy.action.typ = actionBuild
```

Comment:
oh... but why do it this way?

Reply:
That's the least invasive way to stack `options`. Doing it this way keeps the invariant where we enter the procedures doing the action with the correct `typ` set.

Reply:
What do you mean by "stack ``options``", why do you need to do this? 

This looks like a bit of a workaround to me, and you'll undoubtedly get cases where some flags in options are lost because they have not been carried over in ``briefClone``. I just don't understand why this is necessary, why make ``buildFromDir`` expect to only be called for an options with an ``actionBuild``? It doesn't reflect the options that were passed to the application.

Reply:
> This looks like a bit of a workaround to me

It works around a broken design, yes.

> you'll undoubtedly get cases where some flags in options are lost because they have not been carried over in briefClone

I did not introduce this pattern, I've only re-used it for more cases. #495 shows that it was not such a great idea to boot.

>  just don't understand why this is necessary, why make buildFromDir expect to only be called for an options with an actionBuild

You didn't read #533, didn't you? Right now calling `nimble build` propagates the `build` option to every sub-task and that's incredibly wrong if nimble is actually **building** and **installing** some dependencies. The user specified action tells you what to do but should not have any impact on the sub-tasks coordinated by nimble. Another example is the `nimble install` process: the difference between `nimble build` and `nimble install` is that the latter copies some artifacts over to some directory. You can also describe it as `build` + `copy` and it makes sense to propagate `actionBuild` during the first phase and `actionInstall` during the latter because you don't want to duplicate your `task` logic between `build` and `install`.

Reply:
> I did not introduce this pattern, I've only re-used it for more cases. #495 shows that it was not such a great idea to boot.

From what I can tell #495 has nothing to do with how options are handled. It exists simply because the options are not propagated to Nimscript's ``exec``. In fact, it's debatable whether they should be, or whether we should offer an API to retrieve the options and make it easy to pass onto commands.

> Right now calling `nimble build` propagates the `build` option to every sub-task and that's incredibly wrong if nimble is actually **building** and **installing** some dependencies

The ``options`` reflect what Nimble was executed with. Perhaps this is a bad design, but it was created before "tasks" were even conceived.

> Another example is the `nimble install` process: the difference between `nimble build` and `nimble install` is that the latter copies some artifacts over to some directory. You can also describe it as `build` + `copy` and it makes sense to propagate `actionBuild` during the first phase and `actionInstall` during the latter because you don't want to duplicate your `task` logic between `build` and `install`.

Yes, I understand that. But I don't understand what actually changing the options gets you, you're calling the hook for ``actionBuild`` in ``buildFromDir``, I don't see any checks for what kind of options we got. Did you literally just add it as a sanity check?